### PR TITLE
Added method to Product manager for filtering on product attributes.

### PR DESCRIPTION
**Pull request by highbiza**
Hi here is some functionality that allows one to use product attributes more as opposed to subclassing the product model. A method was added to the product manager that allows querying/filtering on product attributes.

example:
```
Product.objects.filter_by_attributes(height__lt=100, google_shopping=True)
```

It will produce queries like this:
```
SELECT "catalogue_product"."id", "catalogue_product"."structure", "catalogue_product"."upc", "catalogue_product"."parent_id", "catalogue_product"."title", "catalogue_product"."slug", "catalogue_product"."description", "catalogue_product"."product_class_id", "catalogue_product"."rating", "catalogue_product"."date_created", "catalogue_product"."date_updated", "catalogue_product"."is_discountable", "catalogue_product"."vat_rate"
    FROM "catalogue_product"
    INNER JOIN "catalogue_productattributevalue"
    ON (
        "catalogue_product"."id" = "catalogue_productattributevalue"."product_id"
    )
    INNER JOIN "catalogue_productattribute"
    ON (
        "catalogue_productattributevalue"."attribute_id" = "catalogue_productattribute"."id"
    )
    INNER JOIN "catalogue_productattributevalue" T4
    ON (
        "catalogue_product"."id" = T4."product_id"
    )
    INNER JOIN "catalogue_productattribute" T5
    ON (
        T4."attribute_id" = T5."id"
    )
    WHERE (
        "catalogue_productattributevalue"."value_integer" < 100
        AND
        "catalogue_productattribute"."code" = height
        AND
        T4."value_boolean" = True
        AND
        T5."code" = google_shopping
    )
    ORDER BY "catalogue_product"."date_created"
DESC
```

### CODE REVIEW

**Comment by specialunderwear** [comment]
Meh I guess this breaks 1.11 compatibility because it uses FilteredRelation

**Comment by solarissmoke** [comment]
Not caring is not an option unfortunately - 1.11 is an LTS release which we have to support until 2021.

**Comment by specialunderwear** [comment]
OK then merge after 2021 :D

**Comment by specialunderwear** [comment]
The best I can do is skip the tests and disable the functionality

**Comment by highbiza** [comment]
No idea what I have to add to make codacy error go away? pylint says this is fine.

**Comment by highbiza** [comment]
ANd this is cheating because this PR was green before :o

**Comment by specialunderwear** [comment]
Whatever you say codecov, last time for me.

**Comment by specialunderwear** [comment]
This thing is nolonger backwards incompatible!

**Comment by specialunderwear** [comment]
Time to merge this?

File: src/oscar/apps/catalogue/managers.py [hunk 1]
```diff
@@ -1,11 +1,81 @@
...
+        num_fields = 0
+        for key, value in filter_kwargs.items():
+            num_fields += 1
+            if LOOKUP_SEP in key:
+                field_name, lookup = key.split(LOOKUP_SEP, 1)
+                self[field_name] = (lookup, value)
+            else:
+                self[key] = (None, value)
+
+        self.num_fields = num_fields
```
Comment (sasha0) [inline src/oscar/apps/catalogue/managers.py]: What is `num_fields` for? It's not used anywhere.

**Comment by sasha0** [comment]
@specialunderwear I've got to review this. Please see comments.

**Comment by specialunderwear** [comment]
ok thanks

**Comment by pjstevns** [comment]
> Not caring is not an option unfortunately - 1.11 is an LTS release which we have to support until 2021.

Python2 was dropped, for very good reasons. It seems perfectly valid to me to drop django-1.11 in the master branch. Anything that helps speed up development and encourage engagement from contributors.

**Comment by fcecchin** [comment]
Is there a way to filter datetime attributes? I've tried with lookup, but it didn't work.
`Product.objects.filter_by_attributes(start_date__lt=timezone.localtime())`

**Comment by specialunderwear** [comment]
> Is there a way to filter datetime attributes? I've tried with lookup, but it didn't work.
> `Product.objects.filter_by_attributes(start_date__lt=timezone.localtime())`

yeah that should work actually. why don't you add the missing test to ``test_product_attributes_query.py``? A date and datetime attribute is allready provided in the fixture.

**Comment by fcecchin** [comment]
> > Is there a way to filter datetime attributes? I've tried with lookup, but it didn't work.
> > `Product.objects.filter_by_attributes(start_date__lt=timezone.localtime())`
> 
> yeah that should work actually. why don't you add the missing test to `test_product_attributes_query.py`? A date and datetime attribute is allready provided in the fixture.

Yeah, that's really works. It's my fault... I will provide tests for this attribute type.

**Pull request merged**